### PR TITLE
Feature/retailer/add product

### PR DESCRIPTION
**Pull request by Kwang-Kian-Hui**

### CODE REVIEW

File: lib/core/infrastructure/firestore_failures.dart [hunk 1]
```diff
@@ -7,6 +7,8 @@ class FirestoreFailures with _$FirestoreFailures {
   const FirestoreFailures._();
   const factory FirestoreFailures.cancelledOperation() = CancelledOperation;
   const factory FirestoreFailures.objectNotFound() = ObjectNotFound;
+  const factory FirestoreFailures.resourceExhausted() = ResourceExhausted;
+  const factory FirestoreFailures.serviceUnavailable() = ServiceUnavailable;
   const factory FirestoreFailures.unknown() = Unknown;
   const factory FirestoreFailures.noConnection() = NoConnection;
 }
```
Comment (sinhaugoh) [inline lib/core/infrastructure/firestore_failures.dart]: too much failure i think.

File: lib/core/presentation/routes/app_router.dart [hunk 2]
```diff
@@ -18,7 +19,9 @@ import 'package:deall/splash/splash_page.dart';
     MaterialRoute(page: SignInPage),
     MaterialRoute(page: SignUpPage),
     MaterialRoute(page: ConsumerSignUpPage),
+    MaterialRoute(page: AddProductPage),
     MaterialRoute(page: RetailerSignUpPage),
+    MaterialRoute(page: AddProductPage),
     MaterialRoute(page: RetailerProfilePage),
```
Comment (sinhaugoh) [inline lib/core/presentation/routes/app_router.dart:25]: duplicate add product page

File: lib/core/shared/providers.dart [hunk 3]
```diff
@@ -19,6 +25,17 @@ final imagePickingRemoteServiceProvider =
...
+// Product providers
+final productListRemoteServiceProvider =
+    Provider((ref) => ProductListRemoteService(ref.watch(firestoreProvider)));
+
+final productListRepoProvider = Provider((ref) =>
+    ProductListRepository(ref.watch(productListRemoteServiceProvider)));
 
+final productListNotifierProvider =
+    StateNotifierProvider<ProductListNotifier, ProductListState>(
+        (ref) => ProductListNotifier(ref.watch(productListRepoProvider)));
```
Comment (sinhaugoh) [inline lib/core/shared/providers.dart]: The product list shouldn't be in core, coz retailer use stream of product, but consumer use future

File: lib/retailer/application/add_product_form_notifier.dart [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        isSaving: true,
+        hasConnection: true,
+      );
+
+      final newProduct = Product(
+        id: '',
+        name: state.name,
+        usualPrice: state.usualPrice,
+        discountedPrice: state.discountedPrice,
+        image: state.imageFile.toString(),
```
Comment (sinhaugoh) [inline lib/retailer/application/add_product_form_notifier.dart]: this shouldn't be working need to use uploadProductImage in imagePicking repo, refer to retailer sign up notifier.

**Review by sinhaugoh — CHANGES_REQUESTED** [review]

**Comment by Kwang-Kian-Hui** [comment]
Changes made, tested and works exactly as it should for products with and without image

File: lib/core/infrastructure/product/product_dto.dart [hunk 5]
```diff
@@ -65,10 +65,22 @@ class ProductDTO with _$ProductDTO {
...
+      'id': id,
+      'name': name,
+      'usualPrice': usualPrice,
+      'discountedPrice': discountedPrice,
+      'image': image,
+      'description': description,
+      'availability': availability,
+      'dateModified': dateModified,
+    };
+  }
```
Comment (sinhaugoh) [inline lib/core/infrastructure/product/product_dto.dart]: dont need to do tojson and fromJson manually, can just remove

File: lib/core/application/value_validator.dart [hunk 6]
```diff
@@ -34,3 +34,11 @@ Either<ValueFailure, Unit> validatePostal(String input) {
...
 Either<ValueFailure, Unit> validateNotEmpty(String input) =>
     input != '' ? right(unit) : left(const ValueFailure.empty());
+
+Either<ValueFailure, Unit> validateUsualPrice(double input) {
+  return input >= 0.0 && input <= 10000 ? right(unit) : left(const ValueFailure.invalidPriceValue());
+}
+
+Either<ValueFailure, Unit> validateDiscountedPrice(double input, double usualPrice) {
+  return input >= 0.0 && input <= 10000 && input < usualPrice ? right(unit) : left(const ValueFailure.invalidPriceValue());
+}
```
Comment (sinhaugoh) [inline lib/core/application/value_validator.dart:50]: for validation of usual price and discounted price u can check if it is empty first, if it is then return empty failure, then proceed to check with invalid value

File: lib/retailer/application/add_product_form_notifier.dart [hunk 7]
```diff
@@ -52,44 +57,91 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
...
+        state.discountedPriceErrorMessage == null) {
       state = state.copyWith(
         isSaving: true,
         hasConnection: true,
       );
 
-      final newProduct = Product(
-        id: '',
+      final uid = _authRepository.getUserId();
+      final pid = _productRepository.getProductId(uid);
```
Comment (sinhaugoh) [inline lib/retailer/application/add_product_form_notifier.dart]: try to use productId instead. make the variable name meaningful

File: lib/retailer/application/add_product_form_notifier.dart [hunk 8]
```diff
@@ -52,44 +57,91 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
...
+            name: state.name,
+            usualPrice: state.usualPrice,
+            discountedPrice: state.discountedPrice,
+            image: filePath,
+            description: state.description,
+            availability: state.availability,
+          );
+        });
+      }
+
```
Comment (sinhaugoh) [inline lib/retailer/application/add_product_form_notifier.dart:143]: can just use newProduct.copyWith(image: filePath)

File: lib/retailer/infrastructure/product_repository.dart [hunk 9]
```diff
@@ -25,24 +26,15 @@ class ProductListRepository {
...
-                  code: 'unavailable',
-                  plugin: "Service is currently unavailable. Please try again later.")
-              .code) {
-        return left(const FirestoreFailures.serviceUnavailable());
-      }
       return left(const FirestoreFailures.unknown());
     }
   }
 
+  String getProductId(String uid){
```
Comment (sinhaugoh) [inline lib/retailer/infrastructure/product_repository.dart]: can rename to generateNewProductId? sounds more meaningful

**Review by sinhaugoh — CHANGES_REQUESTED** [review]
overall looks good, just minor change needed

File: lib/retailer/application/add_product_form_notifier.dart [hunk 10]
```diff
@@ -72,6 +82,18 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
...
+      (valueFailure) => valueFailure.maybeWhen(
+        empty: () {
+          stateCopy = stateCopy.copyWith(
+            discountedPriceErrorMessage: 'This field cannot be empty',
+          );
+        },
+        orElse: () {},
+      ),
+      (r) => stateCopy = stateCopy.copyWith(discountedPriceErrorMessage: null),
+    );
```
Comment (sinhaugoh) [inline lib/retailer/application/add_product_form_notifier.dart]: need to redo this part, logic error

**Review by sinhaugoh — CHANGES_REQUESTED** [review]

**Review by sinhaugoh — APPROVED** [review]
everything correct

**Pull request merged**